cg.1
.\"             -*-Nroff-*-
.\"
.TH CG 1 "29 Aug 2000" "" ""
.SH NAME
cg \- Recursively grep for a pattern and store it.
.SH SYNOPSIS
.B cg 
[ -l ] | [ [ -i ] pattern [ files ] ]
.SH DESCRIPTION
.B "cg "
does a search though text files (usually source code) recursively
for a pattern, storing matches and displaying the output in a
human-readable fashion.  It is intended to give some of the
functionaly of AT&T's cscope(1) tool, with the advantages of
simplicity and not being language-specific.  The script will
colorize output if configured as such.
.sp
It is typically run with a Perl regular expression to search for.
The search can be made case insensitive by using the
.B -i
option.  A list of files may also be specified with an additional
argument after the pattern.  Put the files pattern in quotes to
make it be matched by Perl rather than by the shell.  Running the
script with no arguments will recall the results of the previous
search.  After the search, entries found can be edited using the
vg(1) script.  The
.B -l
option shows the last log made.
.PP
.SH "SOME EXAMPLES"
.B cg
- alone recalls the previous search results.
.sp
.B cg \-i pattern
- search the default list of files for all files matching the pattern
(and case-insensitively).
.sp
.B cg pattern '*.c'
- search recursively for pattern in all *.c files.  This automatically
converts '*' to '.*' and '.' to '\\.' for you and does a Perl pattern
match on all files in the tree.
.sp
.B cg pattern *.c
- search through the shell-expanded list of *.c files, so not done
recursively (in other words, only the files your shell pass to the
script as arguments).
.sp
.B cg \-l
- show the last log made.
.PP
.SH "COMMAND\-LINE OPTIONS"
.TP 0.5i
.B "\-i "
Do a case-insensitive search.
.TP 0.5i
.B "\-l "
Show the last log made.
.TP 0.5i
.B "\-p "
Toggle the default pager option.  
.B cg
has a bulit-in pager function, which can be enabled or disabled by
default (in .cgvgrc).  If the default is enabled, this option disables
the pager; if the default is disabled, this option enables it.
.TP 0.5i
.B "\-P "
Force the built-in pager to be disabled.



.SH FILES
.ne 3
.TP
.I ${HOME}/.cglast
Log file of the last search.
.ne 3
.TP
.I ${HOME}/.cgvgrc
Per-user configuration file (if the defaults are not desireable).
.ne 3
.TP
.I ${HOME}/.cgvg/*
Log files in $HOSTNAME.shell_pid form with the log of the last search.
.fi

.SH "SEE ALSO"
.BR vg (1),
.BR perl (1),
.BR find (1),
.BR grep (1),
.BR cscope (1)

.SH AUTHOR
.B cg
was written by Joshua Uziel <uzi@uzix.org>.


ChangeLog
----------------
Change Log
----------------
August 29, 2000
1.6.1 - Pretty much added two things... 

	* "Colon mode"... which prints out something like this:

41 cgvg-common.pl:94  open (IN, "<$RCFILE");

	  instead of this:

41 cgvg-common.pl  94 open (IN, "<$RCFILE");
	
	  written on the request of a coworker who wanted it to mesh with
	  his editor a bit better.

	* Thanks to the good grace of Jeffrey R. Hay, cgvg now how has
	  built-in pager so that colors will remain while paging and
	  editing can occur while viewing the output.  Check it out. :)

November 3, 1999
1.6.0 - Some decent change to this version.
	
	* Cleaned up the code some.  I've introduced a cgvg-common.pl
	  file.  This does (and will) contain common subroutines and
	  variables.  More stuff will move there as I possibly do a GUI.
	  (If I do one.)  The script are now also broken into subroutines
	  a bit for easier readability and managability.  Some more cleaning
	  has to happen, but this is a start.

	* Logfile storage has changed quite a bit.  Now instead of one
	  global .cglast file, I have a .cgvg/ directory where I store
	  a log file for each shell in the form $HOSTNAME.shell_pid.
	  This is so that a user can use cgvg in several different places
	  and not lose their searches.  The name is for cleanup.  If it
	  doesn't find a pid that there's a log for on the present machine,
	  it deletes the log file.  The last search is still obtainable, but
	  now with a '-l' option... so "cg -l" to see the last search, and
	  "vg -l number" to edit an entry from that log.  Also note that
	  not having a log with the present terminal will default it to
	  the last log.  (Some of this idea and the next two came from Goran
	  Koruga <goran.koruga@hermes.si>)
	
	* There is no more BOLD option.  Now, all colors are achieveable
	  as the color, and b_color for the bold version (so "red" and
	  "b_red" for example).
	
	* The EXCLUDE option is now in the logfile, as is an AGE option
	  for a default cleanup time on logfiles.

	* There is now a configure script for this program to make things
	  (hopefully) more convenient.  (Fix by Manish Singh <yosh@gimp.org>)
	
	* Fix for handling the $EDITOR variable better if the editor is
	  using any arugments.  (Fix by Clinton Popetz <cpopetz@cpopetz.com>)

	* Also searching for *.l, *.y and *.C files by default as well.
	  (Request by Adam Marks <adam@cp.net>)

October 1, 1999
1.5.2 - Just a small bugfix release.  First, I removed a debug statement
	that I forgot to take out before sending off the last version
	(oops).  The bigger change is that now it'll parse through
	directories that are given as arguments... so if you do a
	"cg 'pattern' dir1 dir2", it'll recursively search through
	those directories.

August 25, 1999
1.5.1 - Fixed a bug where if one and only one file was specified
	to search through and it was an existing text file that it
	would read that file rather than blindly assume it's a
	search expression and do a huge recursive search.

	Cleaned up the finding code a bit.  Now both the default
	search and a user-specified search use the same &wanted
	subroutine, with the search list being specified by a
	$SEARCH variable.  Along with this, the default search list
	is user-defineable in the .cgvgrc file as well.

August 18, 1999
1.5 -	The big change of this version is that now the searching is
	handled by Perl rather than by grep(1), egrep(1) and fgrep(1).
	This required quite a bit of change, but everything is working
	like it was before.  The only really desirable option beyond
	the '-n' option (which I've covered) was '-i', which is
	covered as the only option at the moment.  The only things
	that are not handled by Perl now and are hard-coded now are
	"stty" and "which", and when I find a clean way of implementing
	them without requiring a trip to CPAN, I'll add those as well.

	Bugs have been fixed with handling the configuration file,
	and with the user-specified search list.  Now a user can
	search (for example) all C files with "cg pattern *.c" or
	"cg pattern '*.c'"... with the former being expanded by the
	shell and the latter being pattern matched by Perl.  Also
	note that the latter is a recursive search whereas the former
	is only those files specified on the command-line.  Colors
	are also now configurable by the configuration file.

	Manpages and a Makefile have also been added for user convenience.

August 15, 1999
1.4 -	Updates made here fix a bug that not all of the search list
	was being utilized, and the biggest change being that cg
	now uses Perl's internal File::Find routine instead of an
	external /bin/find (or whatever) to get the list of files.

	Syntax has changed a bit because of this... when files to
	search are specified, it must be done in quotes (single or
	double) like "cg pattern '*'" to search all files.

	Also, there's now a ~/.cgvgrc file where users can specify
	their editor, and bold and coloring options so that not all
	system users have to use the default.  See the "cgvgrc_example"
	file for a sample with the defaults.  Currently only $EDITOR,
	$BOLD, $BOLD_ALTERNATE and $COLORS can be set.  A bug in
	version 1.3 was corrected with the $BOLD option as well.

July 7, 1999
1.3 -	Same as the last version, except now colors are supported
	in cg.  There are the options $BOLD, $BOLD_ALTERNATE and  
	$COLOR to be set to 1 (on) or 0 (off) if you do or don't 
	want these feature.  $COLOR is needed for the $BOLD
	options, and really you should select only one or neither
	of the two $BOLD options.  Colors only work when displaying
	straight to STDOUT.

	In vg, I've gotten rid of the $CGEDITOR variable and replaced
	it with the standard $EDITOR environment variable since I've
	found most editors to support the +number option.  Also fixed
	a bug to give a clear error when the editor doesn't exist.

June 30, 1999
1.2 -	More changes to the cg script.  Rather than simply print the
	logfile like previous versions, I actually attempt to make

cg.pl
# cg - Code Grep - grep recursively through files and disiplay matches
# Copyright 1999 by Joshua Uziel <juziel@home.com> - version 1.5.2
#
# usage: cg [-i] [pattern] [files]
#
# Recursive Grep script that does a bit of extra work in adding a count
# field and storing the data in a file, as well as displaying data in a
# colorful and human-readable fashion.  Run with a perl regular expression
# to search for it (with '-i' option for case-insensitive).  You can supply
# a quoted file pattern to search for ('*.c'), run just "cg" alone to
# recall the last search (since it's save to the $LOGFILE), and running
# with a list of files is allowable (though not recurive and pattern-matched
# like the quoted variation).
#
# Examples: "cg printf", "cg printf '*.c'", "cg -i printf '*.c'",
#		"cg -i printf *.c", "cg", etc.
#
# The point of this script was to provide source code searching
# functionality similar to that AT&T's cscope(1).  This is a pure
# hack and lacks any sophistication, but has the advantage that it
# can be used for more than just the C programming language, besides
# adding the functionality that is generally missing from a developer's
# toolbox.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

use POSIX;
use File::Find;
require "find.pl";


# Search for wanted entries for perl internal find subroutine.
sub wanted {

	# Skip things that aren't normal files (like directories).
	if ((($dev,$ino,$mode,$nlink,$uid,$gid) = lstat($_)) && -f _) {

		# Kill the leading ./ and push it on the @LIST
		$name =~ s/^\.\///;

		# Push onto the list if we have a match
		push @LIST, $name if ($name =~ /$SEARCH/o);
	}
}

# Default search list:
# 	Make* *.c *.h *.s *.cc *.pl *.pm *.java *.*sh *.idl
$SEARCH = '(^Make.*$|^.*\.([chs]|cc|p[lm]|java|.*sh|idl)$)';

# Where to store the data
$LOGFILE = "$ENV{'HOME'}/.cglast";

# Path to the rc file
$RCFILE = "$ENV{'HOME'}/.cgvgrc";

# List of files and strings to exclude from our search.
$EXCLUDE = "SCCS|RCS|tags|\.make\.state";

# Set if you want colors (and your term supports it).  This is required
# for the $BOLD* options.
$COLORS = 1;

# Have everything printed in bold... 1 (yes) or 0 (no) only.  This option
# is overrided by $BOLD_ALTERNATE and only available with $COLORS
$BOLD = 0;

# Make every other line bold.
$BOLD_ALTERNATE = 1;

# Defined colors
%colors = ( 'black'	=> "30",
	    'red'	=> "31",
	    'green'	=> "32",
	    'yellow'	=> "33",
	    'blue'	=> "34",
	    'magenta'	=> "35",
	    'cyan'	=> "36",
	    'white'	=> "37");

# Default color for column #
$c[1] = $colors{'cyan'};
$c[2] = $colors{'blue'};
$c[3] = $colors{'red'};
$c[4] = $colors{'green'};

# Check if stdout goes to a tty... can't do colors if we pipe to another
# program like "more" or "less" or to a file.
$COLORS = POSIX::isatty(fileno STDOUT) if ($COLORS);

# If the rc file exists, parse it and override the defaults.
if (-f $RCFILE) {
	open (IN, "<$RCFILE");
	
	while (<IN>) {
		chomp;
		
		# Strip leading spaces and skip blank and comment lines.
		s/\s*//g;
		next if (/^#/);
		next if (/^$/);
		
		($key, $value) = split /=/;

		# Match only the specific value.
		if ($key =~ /^COLORS$/) {
			$COLORS=$value;
		} elsif ($key =~ /^BOLD$/) {
			$BOLD=$value;
		} elsif ($key =~ /^BOLD_ALTERNATE$/) {
			$BOLD_ALTERNATE=$value;
		} elsif ($key =~ /^EDITOR$/) {
			$EDITOR=$value;
		} elsif ($key =~ /^SEARCH$/) {
			$SEARCH=$value;

		# Change colors from the defaults.
		} elsif ($key =~ /^COLOR[1-4]$/) {

			# See that a legal color has been given
			if ($value =~ 
			/^(black|red|green|yellow|blue|magenta|cyan|white)$/) {
				$coltmp = $key;
				$coltmp =~ s/COLOR//;
				$c[$coltmp] = $colors{$value};
			} else {
				die "error: Unknown color '$value' in $RCFILE",
					"at line $..\n";
			}
		} else {
			die "error: Unknown option '$key' in $RCFILE at line",
				"$..\n";
		}
	}

	close (IN);
}

# Generate the log ...
if ($#ARGV+1) {
	$count = 0;
	open (OUT, ">$LOGFILE");

	# Give a point of reference if we change directories.
	print OUT "PWD=$ENV{PWD}\n";

	# Set the @ARGLIST and the file $SEARCH (if any) while counting
	# non-dash arguments.  More than one means we have a $SEARCH, else
	# we use the default list of files to search through.
	$nondash = 0;
	foreach (@ARGV) {
		if (/^-/) {
			push @ARGLIST, $_;
		} else {
			if ($nondash) {
				push @FILELIST, $_;
			} else {
				$pattern = $_
			}
			$nondash++;
		}
	}
	$nondash--;

	# If we have a file list of size 1, use it as a search pattern
	# for files automatically.
	if ($nondash == 1) {
		
		# Unless that one thing is a file, in which case we just
		# search it.
		if (-T $FILELIST[0]) {
			die "error: File $FILELIST[0] not readable.\n"
				unless (-r $FILELIST[0]);
			$nondash++;	# Psych out the $nondash check later.
		} else {
			$SEARCH = $FILELIST[0];
			$SEARCH =~ s/\./\\\./g;		# . --> \.
			$SEARCH =~ s/\*/\.\*/g;		# * --> .*
		}
	}

	# Check our arguments
	$insensitive = 0;
	foreach (@ARGLIST) {
		if (/^\-i$/) {
			$insensitive = 1;
		} else {
			die "error: Unknown argument.\n";
		}
	}

	# Adding "(?i)" to the head makes it case-insensitive
	# (aka. data-driven case insensitivity)
	if ($insensitive) {
		$pattern = "(?i)" . $pattern;
	}

	# Use the given list of files if more than 2 given by the shell,
	# else to a recursive find, matching on the default or given pattern.
	if ($nondash >= 2) {
		@LIST = @FILELIST;
	} else {
		# Initialize @LIST so it's now global and do the find...
		@LIST;
		&find('.');
	}

	# Remove files found in our $EXCLUDE list
	@LIST = grep !/$EXCLUDE/, @LIST;

	# For directories, do the find(s) down that directory.
	for ($i=0; $i<=$#LIST; $i++) {
		if (-d $LIST[$i]) {
			&find($LIST[$i]);
		}
	}

	# Special case of no matching files, we die with an error.
	die "error: No matching files found.\n" if ($#LIST < 0);

	# Search through the list of files and generate the $LOGFILE
	foreach $file (@LIST) {
		# Only open text files (-T) that we can read (-r).
		open(IN, "<$file") if ((-T $file) && (-r $file));

		while (<IN>) {
			# Search for the pattern (o == only compile once)
			if (/$pattern/o) {
				# $. is the line number and $_ is the entry
				print OUT "$count:$file:$.:$_";
				$count++;
			}
		}
		close (IN);
	}
	close (OUT);
}

# Either way, we print the log... this part works to reformat things
# differently from how it's stored to make it easier on the human eyes.

# Attempt to get the number of columns from an "stty -a"
if ($COL = `stty -a | grep column 2> /dev/null`) {

	# Strip out the value with the string "column"
	@TMP = split ';', $COL;
	foreach $tmp (@TMP) {
		$COL = $tmp if ($tmp =~ /column/);
	}
	# Grab the digit characters surrounded by non-digit characters.
	$COL =~ s/\D*(\d+)\D*/$1/;

	# Something's weird if 0, and we want more than 40.
	die "Error: Zero value found for number of columns.\n" if ($COL == 0);
	die "Error: Too few columns to work with.\n" if ($COL < 40);

	# Adjust things to be a little smaller than the width.
	$COL -= 2;
} else {
	# Default assumption is 80 columns, so do 2 less than it.
	$COL = 78;
}

# Exit there's no logfile.
die "Error: $LOGFILE does not exist.\n" unless (-f "$LOGFILE");

open (IN, "<$LOGFILE");
<IN>;	# Waste the first line, used for PWD.

# $m* are used as "max" variables... maximum length at this point.
$mnum = $mline = $mfile = $i = 0;

while ($in = <IN>) {
	chomp $in;

	# Split and strip the first few colons, leave the rest.
	($rec[$i]->{num}, $rec[$i]->{file}, $rec[$i]->{line}, $rec[$i]->{str}) 
		= split /:/, $in, 4;

	# Remove all leading whitespace.
	$rec[$i]->{str} =~ s/^\s*//;

	# Swap tabs for 8 spaces
	$rec[$i]->{str} =~ s/\t/        /g;
	
	# If we have a longer length for this field, save it. 
	$tmp = length $rec[$i]->{file};
	$mfile = $tmp if ($mfile < $tmp);
	$tmp = length $rec[$i]->{line};
	$mline = $tmp if ($mline < $tmp);

	$i++;
}

# Better than doing this every time like $mfile and $line ...
$mnum = length ($i-1);

# Skip inward the 3 lengths and the spaces separating them.
$skip = $mnum + $mfile + $mline + 3;

# Special case I call "wrapmode" when we're to skip so much that we
# can't even fit 20 characters (and in some cases negative characters).
# Go to next line and automatically skip a tab's worth.
if (($skip + 20) >= $COL) {
	$wrapmode = 1;
	$skip = 8;
}

# Length for the string is the whole line minus length of others.
# Hopefully $COL is adjusted terminal's width.
$mstr = $COL - $skip;

$entries = $i;

for ($i=0; $i < $entries; $i++) {

	# Bold every other entry
	$BOLD = ($i % 2) if ($BOLD_ALTERNATE);
	
	# Print the properly justified first 3 fields.
	print "\e[$BOLD;${c[1]}m" if ($COLORS);
	printf "%${mnum}s ", $rec[$i]->{num};  
	print "\e[0m" if ($COLORS);

	print "\e[$BOLD;${c[2]}m" if ($COLORS);
	printf "%-${mfile}s ", $rec[$i]->{file}; 
	print "\e[0m" if ($COLORS);
	
	print "\e[$BOLD;${c[3]}m" if ($COLORS);
	printf "%${mline}s ", $rec[$i]->{line};
	print "\e[0m" if ($COLORS);

	# Newline only for "wrapmode".
	print "\n" if ($wrapmode);

	# Trickery for the string.  Do this as many times as we've got
	# str's length divided by it's maximum possible length.
	for ($j=0; $j < ((length $rec[$i]->{str}) / $mstr); $j++) {

		# Only skip after first line.
		print " " x $skip if ($j || $wrapmode);

		# Print only $mstr character substring.
		print "\e[$BOLD;${c[4]}m" if ($COLORS);
		print substr $rec[$i]->{str}, ($j*$mstr), $mstr;
		print "\e[0m" if ($COLORS);
		print "\n";
	}
}

close (IN);
